### ISSUE TITLE
Feat: Profile Hooks

### ISSUE THREAD

**Issue by Blade67** [better-auth/better-auth#950]
### Is this suited for github?

- [X] Yes, this is suited for github

### Is your feature request related to a problem? Please describe.

It currently seems impossible to update User values based on their linked social provider Profiles, making it hard to sync data between social providers and User tables.

### Describe the solution you'd like

I would suggest Profile hooks that are triggered on each sign-in and potentially sign-out.

```ts
socialProviders: {
    discord: {
        clientId: env.DISCORD_CLIENT_ID || '',
        clientSecret: env.DISCORD_CLIENT_SECRET || '',
        redirectUri:
                '<DiscordOAuthURI>',
        scope: ['identify', 'email'],
        prompt: 'consent',
        mapProfileToUser: (profile) => {
            return {
                image: profile.image_url
            };
        },
        signIn: (user, profile) => {
            return {
                image: user.image !== profile.image_url && profile.image_url // example usage
            }
        },
        signOut: (user, profile) => {    // optional
        },
    }
}
```
In the example above I provide both, the current User as it is saved in the database, and the profile received from the social provider - in this case Discord. This allows for more flexibility.

### Describe alternatives you've considered

One consideration would to be to update the `databaseHooks` session hooks to account for sign-in/out and pass both, Profile and User there too. However this seems a bit too loaded and out of scope for those.

**Comment by Q16solver** [comment]
I need something similar to this, but on a more generic level. I want to have a sign-in hook that lets me update values to the user table which currently doesn’t seem possible

**Comment by Blade67** [comment]
Mutating schema seems a bit tricky. Having some kind of arbitrary data key would likely allow that as well. ie. in PostGres you can have values be arbitrary JSON data. This could be used for your use case, @Q16solver. 
This proposal would mostly allow to modify users upon sign-in/out to begin with, opening a lot of opportunities.

**Comment by Q16solver** [comment]
Woops, just realised the wording there I have is confusing. Extending the core schema is already supported. I meant that I wanted to update values in the schema in the signin hook e.g db.update(lastLogin)

**Comment by arxk** [comment]
Love to have this feature! compatible with Generic OAuth too!

**Comment by prokopsimek** [comment]
I need the same to whitelist e-mail addresses for sign-up/sign-in for social providers too.

**Comment by mfragale** [comment]
True, it does seem impossible to update User values based on their linked social provider Profiles, which is weird because if a user signs up directly using an external social provider their info is imported properly, including their profile image.
